File: app/src/main/java/com/dougkeen/bart/platform/DepartureAlarmScheduler.java
package com.dougkeen.bart.platform;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.util.Log;

import com.dougkeen.bart.model.Constants;
import com.dougkeen.bart.model.Departure;
import com.dougkeen.bart.receivers.AlarmBroadcastReceiver;
import com.dougkeen.util.Observable;

/** Owns Android alarm scheduling for one followed departure. */
public final class DepartureAlarmScheduler implements AutoCloseable {
    private final Context applicationContext;
    private final AlarmManager alarmManager;
    private final Departure departure;
    private final Observable<Integer> leadTimeMinutes = new Observable<>(0);
    private final Observable<Boolean> pending = new Observable<>(false);

    public DepartureAlarmScheduler(Context context, Departure departure) {
        applicationContext = context.getApplicationContext();
        alarmManager = (AlarmManager) applicationContext
                .getSystemService(Context.ALARM_SERVICE);
        this.departure = departure;
    }

    public int getLeadTimeMinutes() {
        return leadTimeMinutes.getValue();
    }

    public Observable<Integer> getLeadTimeMinutesObservable() {
        return leadTimeMinutes;
    }

    public boolean isPending() {
        return pending.getValue();
    }

    public Observable<Boolean> getPendingObservable() {
        return pending;
    }

    public int getSecondsUntilAlarm() {
        return departure.getMeanSecondsLeft() - getLeadTimeMinutes() * 60;
    }

    public void setUp(int leadTimeMinutes) {
        this.leadTimeMinutes.setValue(leadTimeMinutes);
        pending.setValue(true);
        schedule();
    }

    public void update() {
        if (alarmManager == null) {
            Log.w(Constants.TAG, "No alarm manager available, so alarm will not be updated");
            return;
        }
        if (isPending() && getLeadTimeMinutes() > 0) {
            schedule();
        }
    }

    public void cancel() {
        if (alarmManager != null) {
            alarmManager.cancel(getAlarmIntent());
        }
        pending.setValue(false);
        Log.d(Constants.TAG, "Alarm cancelled");
    }

    public void notifyAlarmHasBeenHandled() {
        pending.setValue(false);
    }

    @Override
    public void close() {
        if (isPending()) {
            cancel();
        }
        leadTimeMinutes.unregisterAllObservers();
        pending.unregisterAllObservers();
    }

    private PendingIntent getAlarmIntent() {
        Intent intent = new Intent(applicationContext, AlarmBroadcastReceiver.class);
        intent.setAction(Constants.ACTION_ALARM);
        return PendingIntent.getBroadcast(applicationContext, 0, intent,
                PendingIntent.FLAG_IMMUTABLE);
    }

    private long getAlarmClockTime() {
        return departure.getMeanEstimate() - getLeadTimeMinutes() * 60 * 1000L;
    }

    private void schedule() {
        if (alarmManager == null) {
            Log.w(Constants.TAG, "No alarm manager available, so alarm will not be scheduled");
            return;
        }

        long alarmTime = getAlarmClockTime();
        PendingIntent alarmIntent = getAlarmIntent();
        if (alarmTime < System.currentTimeMillis()) {
            alarmManager.set(AlarmManager.RTC_WAKEUP, alarmTime, alarmIntent);
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S
                    && !alarmManager.canScheduleExactAlarms()) {
                alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, alarmTime,
                        alarmIntent);
                Log.w(Constants.TAG,
                        "Exact alarm permission is unavailable; using an inexact alarm");
                return;
            }
            try {
                alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, alarmTime,
                        alarmIntent);
            } catch (SecurityException exception) {
                alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, alarmTime,
                        alarmIntent);
                Log.w(Constants.TAG,
                        "Exact alarm permission is unavailable; using an inexact alarm");
            }
        } else {
            try {
                alarmManager.setExact(AlarmManager.RTC_WAKEUP, alarmTime, alarmIntent);
            } catch (SecurityException exception) {
                alarmManager.set(AlarmManager.RTC_WAKEUP, alarmTime, alarmIntent);
                Log.w(Constants.TAG,
                        "Exact alarm permission is unavailable; using a regular alarm");
            }
        }

        Log.v(Constants.TAG, "Scheduling alarm for "
                + android.text.format.DateFormat.format("h:mm:ss", alarmTime));
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/platform/DepartureAlarmScheduler.kt
package com.dougkeen.bart.platform

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.core.content.edit
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale
import com.dougkeen.bart.model.Constants
import com.dougkeen.bart.model.Departure
import com.dougkeen.bart.model.Station
import com.dougkeen.bart.model.SystemTimeSource
import com.dougkeen.bart.model.TimeSource
import com.dougkeen.bart.receivers.AlarmBroadcastReceiver
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Owns Android alarm scheduling for one followed departure. */
class DepartureAlarmScheduler @JvmOverloads constructor(
    context: Context,
    private val departure: Departure,
    private val timeSource: TimeSource = SystemTimeSource
) :
    AutoCloseable {

    private companion object {
        const val ALARM_PREFS = "departure_alarm_state"
        const val LEAD_TIME_SUFFIX = ".leadTimeMinutes"
        const val PENDING_SUFFIX = ".pending"
    }

    private val applicationContext = context.applicationContext
    private val alarmManager = applicationContext
        .getSystemService(Context.ALARM_SERVICE) as? AlarmManager
    private val preferences = applicationContext.getSharedPreferences(
        ALARM_PREFS, Context.MODE_PRIVATE)
    private val stateKey = buildStateKey(departure)
    private val _state = MutableStateFlow(
        DepartureAlarmState(
            leadTimeMinutes = preferences.getInt(
                stateKey + LEAD_TIME_SUFFIX, 0),
            pending = preferences.getBoolean(stateKey + PENDING_SUFFIX, false),
        )
    )
    val state: StateFlow<DepartureAlarmState> = _state.asStateFlow()

    init {
        val nowMillis = timeSource.nowMillis()
        if (DepartureAlarmPolicy.shouldRestore(isPending,
                departure.hasDeparted(nowMillis), departure.hasExpired(nowMillis))) {
            schedule()
        } else if (isPending) {
            cancel()
        }
    }

    val leadTimeMinutes: Int
        get() = _state.value.leadTimeMinutes

    val isPending: Boolean
        get() = _state.value.pending

    val secondsUntilAlarm: Int
        get() = DepartureAlarmPolicy.secondsUntilAlarm(
            departure.getMeanEstimate(), leadTimeMinutes, timeSource.nowMillis())

    fun setUp(leadTimeMinutes: Int) {
        require(leadTimeMinutes >= 0) {
            "leadTimeMinutes must be non-negative"
        }
        updateState(leadTimeMinutes, true)
        schedule()
    }

    fun cancel() {
        alarmManager?.cancel(alarmIntent())
        updateState(leadTimeMinutes, false)
        Log.d(Constants.TAG, "Alarm cancelled")
    }

    fun notifyAlarmHasBeenHandled() {
        updateState(leadTimeMinutes, false)
    }

    /**
     * Closes this scheduler. When a live feed replaces the Departure instance for
     * the same train, the pending preference must survive long enough for the new
     * scheduler to restore and reschedule it with the updated ETA.
     */
    fun close(preservePending: Boolean = false) {
        if (isPending) {
            alarmManager?.cancel(alarmIntent())
            if (!preservePending) {
                updateState(leadTimeMinutes, false)
            }
        }
    }

    override fun close() {
        close(preservePending = false)
    }

    private fun alarmIntent(): PendingIntent {
        val intent = Intent(applicationContext, AlarmBroadcastReceiver::class.java)
            .setAction(Constants.ACTION_ALARM)
        return PendingIntent.getBroadcast(
            applicationContext,
            0,
            intent,
            PendingIntent.FLAG_IMMUTABLE,
        )
    }

    private fun alarmClockTime(): Long = DepartureAlarmPolicy.alarmTime(
        departure.getMeanEstimate(), leadTimeMinutes)

    private fun schedule() {
        val manager = alarmManager
        if (manager == null) {
            Log.w(Constants.TAG,
                "No alarm manager available, so alarm will not be scheduled")
            return
        }

        val alarmTime = alarmClockTime()
        val intent = alarmIntent()
        if (alarmTime < timeSource.nowMillis()) {
            manager.set(AlarmManager.RTC_WAKEUP, alarmTime, intent)
        } else if (!manager.canScheduleExactAlarms()) {
            manager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, alarmTime, intent)
            Log.w(Constants.TAG,
                "Exact alarm permission is unavailable; using an inexact alarm")
        } else {
            try {
                manager.setExactAndAllowWhileIdle(
                    AlarmManager.RTC_WAKEUP, alarmTime, intent)
            } catch (exception: SecurityException) {
                manager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, alarmTime, intent)
                Log.w(Constants.TAG,
                    "Exact alarm permission is unavailable; using an inexact alarm")
            }
        }

        val alarmText = DateTimeFormatter.ofLocalizedTime(FormatStyle.MEDIUM)
            .withLocale(Locale.getDefault())
            .withZone(ZoneId.systemDefault())
            .format(Instant.ofEpochMilli(alarmTime))
        Log.v(Constants.TAG, "Scheduling alarm for $alarmText")
    }

    private fun updateState(leadTimeMinutes: Int, pending: Boolean) {
        _state.value = DepartureAlarmState(leadTimeMinutes, pending)
        preferences.edit {
            putInt(stateKey + LEAD_TIME_SUFFIX, leadTimeMinutes)
            putBoolean(stateKey + PENDING_SUFFIX, pending)
        }
    }

    private fun buildStateKey(departure: Departure): String = buildString {
        append("alarm.")
        appendStation(this, departure.origin)
        appendStation(this, departure.trainDestination)
        append('|').append(departure.line)
        append('|').append(departure.direction)
        append('|').append(departure.platform)
    }

    private fun appendStation(builder: StringBuilder, station: Station?) {
        builder.append('|').append(station?.abbreviation ?: "")
    }
}
